### ISSUE TITLE
DNS definition by Kubernetes affects alertmanager

### ISSUE THREAD

**Issue by oded-dd** [coreos/prometheus-operator#2934]
**What happened?**
Alertmanager `cluster.peer` (added by prometheus-operator) does not translate as a fully qualified name but as:

```
alertmanager-central-alertmanager-0.alertmanager-operated.default.svc.default.svc.cluster.local SOA ns.dns.cluster.local
alertmanager-central-alertmanager-0.alertmanager-operated.default.svc.svc.cluster.local SOA ns.dns.cluster.local
```

DNS resolution inside a container - like any Linux system - is driven by the /etc/resolv.conf config file. The default Kubernetes dnsPolicy is `ClusterFirst` which means any DNS query will be routed to dnsmasq running in the kube-dns pod inside the cluster which - in turn - will route the request to kube-dns application if the name ends with a cluster suffix or to the upstream DNS server otherwise.

The /etc/resolv.conf file inside each container will look like this by default:
```
nameserver xxx.xxx.xxx.xxx
search namespace.svc.cluster.local svc.cluster.local cluster.local eu-central-1.compute.internal
options ndots:5
```

A fully qualified name is a name for which no local search will be executed and the name will be treated as an absolute one during the resolution. Prometheus-operator should add it like         `- --cluster.peer=alertmanager-central-alertmanager-0.alertmanager-operated.default.svc.:9094`
instead of 
`- --cluster.peer=alertmanager-central-alertmanager-0.alertmanager-operated.default.svc:9094`
(dot was added to the end, in order to treat it as FQDN)

**Did you expect to see some different?**
I would expect it to resolve as a fully qualified name

* Prometheus Operator version:
0.32

* Kubernetes version information:
1.15.6

* Kubernetes cluster kind:
kops

* Manifests:
```
insert manifests relevant to the issue
Alertmanager statefulset created by the `prometheus-operator`
```
    spec:
      containers:
      - args:
        - --config.file=/etc/alertmanager/config/alertmanager.yaml
        - --cluster.listen-address=[$(POD_IP)]:9094
        - --storage.path=/alertmanager
        - --data.retention=744h
        - --web.listen-address=:9093
        - --web.route-prefix=/
        - --log.format=json
        - --cluster.peer=alertmanager-central-alertmanager-0.alertmanager-operated.default.svc:9094
```

**Comment by brancz** [comment]
I wouldn’t consider this a bug but rather an enhancement, but feel free to make this change, if all tests pass I’m happy to merge this.

**Comment by vukasin-nikodijevic** [comment]
I see this as bug as `cluster.local` is missing at the end domain. 

Alertmanager is throwing dns resolution errors:
```
alertmanager-prometheus-operator-alertmanager-0 alertmanager level=warn ts=2020-02-14T12:38:53.143Z caller=cluster.go:228 component=cluster msg="failed to join cluster" err="1 error occurred:\n\t* Failed to resolve alertmanager-prometheus-operator-alertmanager-0.alertmanager-operated.monitoring.svc:9094: lookup alertmanager-prometheus-operator-alertmanager-0.alertmanager-operated.monitoring.svc on 10.0.0.10:53: no such host\n\n"
```

This is not matter of config (it should be) but hardcoded:
https://github.com/coreos/prometheus-operator/blob/master/pkg/alertmanager/statefulset.go#L327

I'll expect to see something like this:
```
  - args:
.
.
.
    - --cluster.peer=alertmanager-prometheus-operator-alertmanager-0.alertmanager-operated.monitoring.svc.cluster.local:9094
``` 

Or even better parametrise `cluster.local` and set as default value to the parameter.

### From the alertmanager pod:

```
$ k exec -it  alertmanager-prometheus-operator-alertmanager-0 sh                                                                              
Defaulting container name to alertmanager.
Use 'kubectl describe pod/alertmanager-prometheus-operator-alertmanager-0 -n monitoring' to see all of the containers in this pod.

/alertmanager $ nslookup alertmanager-prometheus-operator-alertmanager-0.alertmanager-operated.monitoring.svc
Server:		10.0.0.10
Address:	10.0.0.10:53

** server can't find alertmanager-prometheus-operator-alertmanager-0.alertmanager-operated.monitoring.svc: NXDOMAIN

/alertmanager $ nslookup alertmanager-prometheus-operator-alertmanager-0.alertmanager-operated.monitoring.svc.cluster.local
Server:		10.0.0.10
Address:	10.0.0.10:53

Name:	alertmanager-prometheus-operator-alertmanager-0.alertmanager-operated.monitoring.svc.cluster.local
Address: 10.220.1.168

/alertmanager $ 
```

**Comment by brancz** [comment]
Happy to have a field to configure cluster extension.

**Comment by SerialVelocity** [comment]
I just hit this as well.

Every 15s each alertmanager triggers 6 DNS requests:
```
2020-04-11T08:22:30.603502588Z [INFO] 172.30.21.131:36349 - 37012 "A IN alertmanager-alertmanager-0.alertmanager-operated.monitoring.svc.monitoring.svc.cluster.internal. udp 117 false 512" NXDOMAIN qr,aa,rd 228 0.000576357s
2020-04-11T08:22:30.603697746Z [INFO] 172.30.21.131:59789 - 61849 "AAAA IN alertmanager-alertmanager-0.alertmanager-operated.monitoring.svc.monitoring.svc.cluster.internal. udp 117 false 512" NXDOMAIN qr,aa,rd 228 0.000476498s
2020-04-11T08:22:30.607859671Z [INFO] 172.30.21.131:44336 - 24256 "AAAA IN alertmanager-alertmanager-0.alertmanager-operated.monitoring.svc.svc.k8s.internal. udp 106 false 512" NXDOMAIN qr,aa,rd 217 0.000915355s
2020-04-11T08:22:30.610225239Z [INFO] 172.30.21.131:50332 - 56058 "A IN alertmanager-alertmanager-0.alertmanager-operated.monitoring.svc.svc.k8s.internal. udp 106 false 512" NXDOMAIN qr,aa,rd 217 0.001672081s
2020-04-11T08:22:30.617675302Z [INFO] 172.30.21.131:51769 - 37322 "AAAA IN alertmanager-alertmanager-0.alertmanager-operated.monitoring.svc.k8s.internal. udp 102 false 512" NOERROR qr,aa,rd 213 0.000888515s
2020-04-11T08:22:30.617815927Z [INFO] 172.30.21.131:45219 - 10161 "A IN alertmanager-alertmanager-0.alertmanager-operated.monitoring.svc.k8s.internal. udp 102 false 512" NOERROR qr,aa,rd 202 0.003653045s
```

@vukasin-nikodijevic I think you have something wrong in your set up as Kubernetes uses `ndots:5` by default so the domain name should be automatically expanded and there should be no DNS error.

Also, `cluster.local` should definitely be configurable as lots of companies (and some personal stacks like mine) configure this to not be the default.

Another thing that should probably be done is shorten this to `alertmanager-alertmanager-0.alertmanager-operated` so that when the cluster domain flag is not specified, it is only one DNS lookup.
